### PR TITLE
Replace the to_str() by the operator<<

### PR DESCRIPTION
**Pull request by guilhermeferreira**
The insertion operator (<<) is the standard way C++ handles output.
The to_str() is a Java way of outputing string.

Signed-off-by: Guilherme Maciel Ferreira <guilherme.maciel.ferreira@gmail.com>

### CODE REVIEW

**Comment by fpagliughi** [comment]
I'm on the fence with this one, and I admit my thinking is scattered and the existing implementation is lacking. We should remember that, although a `message` payload can be held in a `std::string`, it may actually be a binary blob. So a stream inserter might be misleading, but for a `to_str()` method to be completely useful it would need to be `virtual` so that a user could inherit and override, if they wanted to convert a blob to something printable.

And then we should consider the use of the feature, whichever way it is implemented. Yes, an ostream inserter is the standard way to output a C++ value to a stream, but there are several other reasons to want a string representation of an object, and I always hate having to create and tear down a stringstream just to get a string. It's pretty inefficient - at least in older GCC versions.

So I could probably be persuaded to add ostream inserters, but I doubt I would remove the to_str() methods. I should fix them, though.

File: src/message.cpp [hunk 1]
```diff
@@ -131,3 +131,14 @@ void message::set_payload(const std::string& payload)
 // end namespace mqtt
 }
 
+std::ostream& operator<<(std::ostream& os, const mqtt::message& msg)
+{
+	os << std::string(msg.get_payload());
```
Comment (guilhermeferreira) [inline src/message.cpp]: Payload might be binary.

File: src/message.cpp [hunk 2]
```diff
@@ -131,3 +131,14 @@ void message::set_payload(const std::string& payload)
...
+	return os;
+}
+
+std::ostream& operator<<(std::ostream& os, const mqtt::message_ptr msg)
+{
+	os << std::string(msg->get_payload());
```
Comment (guilhermeferreira) [inline src/message.cpp]: Payload might be binary.

File: src/mqtt/connect_options.h [hunk 3]
```diff
@@ -183,7 +183,6 @@ class connect_options
 	 */
 	/*protected*/ void set_will(const std::string& top, message msg, int qos, bool retained);
 
-	std::string to_str() const;
```
Comment (guilhermeferreira) [inline src/mqtt/connect_options.h]: Keep this method, but qualify as virtual.

File: src/mqtt/exception.h [hunk 4]
```diff
@@ -58,11 +58,6 @@ class exception : public std::runtime_error
 	 */
 	int get_reason_code() const { return code_; }
 	/**
-	 * Returns a string representation of this exception. 
-	 * @return A string representation of this exception. 
-	 */
-	std::string to_str() const { return std::string(what()); }
```
Comment (guilhermeferreira) [inline src/mqtt/exception.h]: Keep this method, but qualify as virtual.

File: src/mqtt/message.h [hunk 5]
```diff
@@ -169,11 +169,6 @@ class message
 	 */
 	void set_retained(bool retained) { msg_.retained = (retained) ? (!0) : 0; }
 	/**
-	 * Returns a string representation of this messages payload. 
-	 * @return std::string 
-	 */
-	std::string to_str() const { return get_payload(); }
```
Comment (guilhermeferreira) [inline src/mqtt/message.h]: Keep this method, but qualify as virtual.

File: src/mqtt/topic.h [hunk 6]
```diff
@@ -95,11 +95,6 @@ class topic
 	 * @return delivery_token 
 	 */
 	idelivery_token_ptr publish(message_ptr msg);
-	/**
-	 * Returns a string representation of this topic.
-	 * @return std::string 
-	 */
-	std::string to_str() const { return name_; }
```
Comment (guilhermeferreira) [inline src/mqtt/topic.h]: Keep this method, but qualify as virtual.

File: src/samples/async_subscribe.cpp [hunk 7]
```diff
@@ -116,7 +116,7 @@ class callback : public virtual mqtt::callback,
 	virtual void message_arrived(const std::string& topic, mqtt::message_ptr msg) {
 		std::cout << "Message arrived" << std::endl;
 		std::cout << "\ttopic: '" << topic << "'" << std::endl;
-		std::cout << "\t'" << msg->to_str() << "'\n" << std::endl;
```
Comment (guilhermeferreira) [inline src/samples/async_subscribe.cpp]: Keep this method, but qualify as virtual.

**Comment by guilhermeferreira** [comment]
Hi Frank. Now I understand the motivation behind the `to_str()` method. Personally I don't think it's worth add more code if it is not used. Then I'll keep this branch in stand by.

**Pull request closed without merging**